relay-cogs/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+//!         match value {
+//!             Message::Span => AppFeatures::new(AppFeature::Spans),
+//!             Message::Transaction => AppFeatures::new(AppFeature::Transactions),
+//!             Message::TransactionWithSpans { num_spans } => AppFeatures::builder()
+//!                 .weight(AppFeature::Spans, *num_spans)
+//!                 .weight(AppFeature::Transactions, 1)
+//!                 .build(),
+//!         }
+//!     }
+//! }
+//!
+//! impl Processor {
+//!     fn handle_message(&self, mut message: Message) {
+//!         let _cogs = self.cogs.timed(ResourceId::Relay, &message);
+//!
+//!         self.step1(&mut message);
+//!         self.step2(&mut message);
+//!
+//!         // Measurement automatically recorded here.
+//!     }
+//! #   fn step1(&self, _: &mut Message) {}
+//! #   fn step2(&self, _: &mut Message) {}
+//! }
+//! ```
+#![warn(missing_docs)]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png",
+    html_favicon_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png"
+)]
+
+use std::time::Duration;
+
+mod cogs;
+mod recorder;
+#[cfg(test)]
+mod test;
+mod utils;
+
+pub use cogs::*;
+pub use recorder::*;
+#[cfg(test)]
+pub use test::*;
+
+/// Resource ID as tracked in COGS.
+///
+/// Infrastructure costs are labeled with a resource id,
+/// these costs need to be broken down further by the application
+/// by [app features](AppFeature).
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum ResourceId {
+    /// The Relay resource.
+    ///
+    /// This includes all computational costs required for running Relay.
+    Relay,
+}
+
+/// App feature a COGS measurement is related to.
+///
+/// App features break down the cost of a [`ResourceId`], the
+/// app features do no need to directly match a Sentry product.
+/// Multiple app features are later grouped and aggregated to determine
+/// the cost of a product.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum AppFeature {
+    /// A placeholder which should not be emitted but can be emitted in rare cases,
+    /// for example error scenarios.
+    ///
+    /// It can be useful to start a COGS measurement before it is known
+    /// what the measurement should be attributed to.
+    /// For example when parsing data, the measurement should be started
+    /// before parsing, but only after parsing it is known what to attribute
+    /// the measurement to.
+    Unattributed,
+
+    /// Metrics are attributed by their namespace, whenever this is not possible
+    /// or feasible, this app feature is emitted instead.
+    UnattributedMetrics,
+    /// When processing an envelope cannot be attributed or is not feasible to be attributed
+    /// to a more specific category, this app feature is emitted instead.
+    UnattributedEnvelope,
+
+    /// Transactions.
+    Transactions,
+    /// Errors.
+    Errors,
+    /// Spans.
```

Comment:
Is this granularity is enough ?  What I mean is e.g. transactions can also contain profiles  or transactions and errors can also contain attachments - do we want to have separate `AppFeature` for those as well?

Reply:
Probably makes sense at some point, I'll make some issues with improvements. 

For now these groups follow what we did before manually and hopefully the results match...